lib/latte.ts

```diff
@@ -0,0 +1,482 @@
+
+    // short circuit when the suite doesn't actually change
+    if (context && context._suite === nextSuite) return context
+
+    // figure out what the common ancestor is between both suites. If either suite is null
+    // (happens at beginning and end of run) then commonAncestor will be undefined, and our
+    // slice calls below will yield the full lineage.
+    const currLineage = lineage(currentSuite)
+    const nextLineage = lineage(nextSuite)
+    const commonAncestor = currLineage.filter(
+      (x) => nextLineage.indexOf(x) >= 0
+    )[0]
+
+    // walk the lineage up to (but not including) the common ancestor, running after callbacks
+    let currTop : number | undefined = currLineage.indexOf(commonAncestor)
+    currTop = currTop == -1 ? undefined : currTop
+    let chain = currLineage.slice(0, currTop)
+    for (const suite of chain) {
+      for (const cb of suite.after) await runWithTimeout(cb, context)
+      context = Object.getPrototypeOf(context)
+    }
+
+    // now walk down the lineage from right below the common ancestor to the new suite, running before callbacks
+    let nextTop : number | undefined = nextLineage.indexOf(commonAncestor)
+    nextTop = nextTop == -1 ? undefined : nextTop
+    chain = nextLineage.slice(0, nextTop).reverse()
+    for (const suite of chain) {
+      context = Object.create(context)
+      context._suite = suite
+      // eslint-disable-next-line @typescript-eslint/no-empty-function
+      context.timeout = function () {}
+      attachHelpers(suite, context)
+      for (const cb of suite.before) await runWithTimeout(cb, context)
+    }
+
+    return context
+  }
+
+  // Run user code with a timeout
+  async function runWithTimeout(cbOrTest: Test | TestSetupCb, context: TestContext) {
+    let hasFinished = false
+    let timeoutPromise: Promise<unknown> | undefined = undefined
+    const setTimeoutPromise = (ms: number) => {
+      timeoutPromise = new RealPromise((res, rej) => {
+        currentTimeout = realSetTimeout(() => {
+          if (hasFinished) return
+          console.error('Timeout', cbOrTest.stack)
+          if (mode == 'headless') {
+            rej(
+              `Timeout, the test took more than ${ms / 1000}s on remote: ${
+                cbOrTest.stack
+              }`
+            )
+          }
+        }, ms)
+      })
+    }
+    setTimeoutPromise(10000)
+
+    context.zen = {}
+    context.zen.extendRemoteTimeout = (ms: number) => {
+      clearTimeout(currentTimeout)
+      setTimeoutPromise(ms)
+    }
+
+    let runPromise = null
+    // TODO this is probably the point where 'call' is coming from
+    if (!cbOrTest.fn) throw new Error('No valid test')
+
+    if (cbOrTest.fn.length > 0) {
+      runPromise = new RealPromise((res) => cbOrTest.fn?.call(context, res))
+    } else {
+      runPromise = cbOrTest.fn.call(context)
+    }
+
+    const race = RealPromise.race([timeoutPromise, runPromise])
+    return promiseFinally.call(race, () => {
```

Comment:
Nit: if we only care about chrome, we can drop the polyfill, and just `race.finally`
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/finally